Dockerize and publish to Quay

rajadain opened the pull request:
## Overview

Instead of building a JAR which gets included in a Docker image by the client repository, we publish a Docker image on Quay which can immediately be included by the client.

This branch is a cleaned up version of work done in [feature/hmc/docker](https://github.com/azavea/usace-program-analysis-geoprocessing/tree/feature/hmc/docker).
## Testing Instructions

Test Docker by running

```
$ ./sbt "project geop" assembly
$ docker build -t usace-geop-test .
$ docker run --rm --publish 127.0.0.1:8090:8090 usace-geop-test
$ curl http://localhost:8090
OK
```

Not sure what the best way is to test Travis, but one way is to see success in the [final build of the source branch](https://travis-ci.org/azavea/usace-program-analysis-geoprocessing/builds/135153748), and the diff between this work and the source branch:

`````` diff
diff --git a/.gitignore b/.gitignore
index 6d907a5..c3b4ce1 100644
--- a/.gitignore
+++ b/.gitignore
@@ -19,6 +19,3 @@ project/plugins/project/
 .scala_dependencies
 .worksheet
 .idea
-
-# CI Specific
-docker/*.jar
diff --git a/.travis.yml b/.travis.yml
index b61fdca..35e561b 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -29,12 +29,14 @@ deploy:
     on:
       repo: azavea/usace-program-analysis-geoprocessing
       tags: true
+
   - provider: script
     script: "scripts/cibuild.sh"
     skip_cleanup: true
     on:
       repo: azavea/usace-program-analysis-geoprocessing
-      branch: feature/hmc/docker
+      branch: develop
+
   - provider: script
     script: "scripts/cibuild.sh"
     skip_cleanup: true
diff --git a/README.md b/README.md
index 207dcd4..a1a1fd8 100644
--- a/README.md
+++ b/README.md
@@ -49,12 +49,13 @@ To test with the [main web app](https://github.com/azavea/usace-program-analysis

 ## Deploy

-Deployments to GitHub Releases are handled via [Travis-CI](https://travis-ci.org/azavea/usace-program-analysis-geoprocessing). The following `git-flow` commands signal to Travis that we want to create a release. The `version` variable should be updated in `project/Version.scala.`
+Deployments to [GitHub Releases](https://github.com/azavea/usace-program-analysis-geoprocessing/releases) and [Quay](https://quay.io/repository/usace/program-analysis-geoprocessing) are handled via [Travis-CI](https://travis-ci.org/azavea/usace-program-analysis-geoprocessing). The following `git-flow` commands signal to Travis that we want to create a release. The `version` variable should be updated in `project/Version.scala` and `Dockerfile`.

 ``` bash
 $ git flow release start 0.0.1
 $ vim CHANGELOG.md
 $ vim project/Version.scala
+$ vim Dockerfile
 $ git commit -m "0.0.1"
 $ git flow release publish 0.0.1
 $ git flow release finish 0.0.1
``````

Connects https://github.com/azavea/usace-program-analysis/issues/5

rajadain committed:
Dockerize and publish to Quay

Instead of building a JAR which gets included in a Docker image by
the client repository, we publish a Docker image on Quay which can
immediately be included by the client.

hectcastro commented:
It looks like this may actually be valid `.dockerignore` syntax. Adding this to the directory that contains the `Dockerfile` should reduce the amount of local data sent to the Docker daemon for the build context:

```
*
!geop/target/scala-2.10/*.jar
```

Commit:
Reduce local data sent to Docker daemon

rajadain commented:
That's a great find, I added it.

On Friday I tried moving the Dockerfile to a folder with nothing else, but found that too confusing and not in keeping with our convention of keeping the Dockerfile in the root. This is much nicer.

hectcastro commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+
+WORKDIR /opt/geoprocessing
+
+EXPOSE 8090
```
I think that we need to introduce a `VOLUME` directive here to ensure that Spark's scratch space gets its own volume:

``` dockerfile
VOLUME /tmp
```

With a volume for `/tmp`:

``` bash
$ docker diff 9899                                                    
```

Without a volume for `/tmp`:

``` bash
❯ docker diff 8f9                                                     
C /tmp
A /tmp/blockmgr-004afef6-fd3d-4aa7-bbfd-1288eea028a9
C /tmp/hsperfdata_root
A /tmp/hsperfdata_root/1
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105/httpd-79c9aff5-aadf-45cc-ab4d-9dc5edfabcb1
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105/httpd-79c9aff5-aadf-45cc-ab4d-9dc5edfabcb1/files
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105/httpd-79c9aff5-aadf-45cc-ab4d-9dc5edfabcb1/jars
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105/httpd-79c9aff5-aadf-45cc-ab4d-9dc5edfabcb1/jars/usace-programanalysis-geop-assembly-0.1.0.jar
A /tmp/spark-3da5844d-5755-48cf-8e9b-6987b7fb2105/userFiles-bd745ca7-c2e3-4849-a878-e1d42ad999fc
```

hectcastro commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+
+EXPOSE 8090
+
+ENTRYPOINT ["/bin/bash", "-c", "spark-submit /opt/geoprocessing/usace-programanalysis-geop-assembly-${VERSION}.jar"]
```
We may want to change this to an `ENTRYPOINT` script so that we can more easily override options passed to `spark-submit` at container launch time:

``` diff
$ git diff Dockerfile         
diff --git a/Dockerfile b/Dockerfile
index 68d8548..26805c3 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -2,10 +2,11 @@ FROM quay.io/azavea/spark:1.6.1

 ENV VERSION 0.1.0

-COPY geop/target/scala-2.10/usace-programanalysis-geop-assembly-${VERSION}.jar /opt/geoprocessing/
+COPY geop/target/scala-2.10/usace-programanalysis-geop-assembly-${VERSION}.jar /opt/geoprocessing/usace-programanalysis-geop.jar
+COPY scripts/docker-entrypoint.sh /opt/geoprocessing/

 WORKDIR /opt/geoprocessing

 EXPOSE 8090

-ENTRYPOINT ["/bin/bash", "-c", "spark-submit /opt/geoprocessing/usace-programanalysis-geop-assembly-${VERSION}.jar"]
+ENTRYPOINT ["./docker-entrypoint.sh"]
$ cat scripts/docker-entrypoint.sh 
#!/usr/bin/env bash

set -e

exec /opt/spark/bin/spark-submit "$@" usace-programanalysis-geop.jar 2>&1
```

This way we can launch the container image with a larger amount of driver memory, for example:

``` bash
$ docker run --rm -p 4040:4040 quay.io/usace/program-analysis-geoprocessing:f98d17f --driver-memory 2G
```

Commit:
Bypass /tmp while layering the container

Spark writes temporary files to /tmp which by default change the
state of the container. By specifying they be mounted, we ensure
that the files under /tmp do not participate in the layering of
the container, and don't show up under `docker diff`.

Commit:
Allow passing arguments to spark-submit

This new approach allows one to pass arguments when running the
container to spark-submit, allowing one to specify operational
parameters at runtime.

rajadain commented:
I think the most recent commits should capture everything.

I looked into using `program-analysis` instead of `programanalysis` as the package name in Scala, so we can have consistency between the repo name `usace-program-analysis-geoprocessing` and the JAR file `usace-programanalysis-geop` but having hyphens there is generally frowned upon. Between having the JAR file match the repo name (run / build time) and the class name (design time), I preferred it matching the class name, since now with Docker the actual JAR file is pretty much abstracted away for run time, and build time values are probably not going to be touched very much after this.

hectcastro commented:
👍 Nicely done.

rajadain commented:
Thanks for pairing with me on this and teaching me about Docker and Quay!
